src/main/java/uk/gov/hmcts/reform/sandl/snlevents/actions/hearing/VacateHearingAction.java

```diff
@@ -0,0 +1,150 @@
+    protected HearingPartRepository hearingPartRepository;
+    protected StatusConfigService statusConfigService;
+    protected StatusServiceManager statusServiceManager;
+    protected EntityManager entityManager;
+
+    // id & hearing part string
+    private Map<UUID, String> originalHearingParts;
+    private String previousHearing;
+    private UserTransactionDataPreparerService utdps = new UserTransactionDataPreparerService();
+
+    public VacateHearingAction(
+        VacateHearingRequest vacateHearingRequest,
+        HearingRepository hearingRepository,
+        HearingPartRepository hearingPartRepository,
+        StatusConfigService statusConfigService,
+        StatusServiceManager statusServiceManager,
+        ObjectMapper objectMapper,
+        EntityManager entityManager
+    ) {
+        this.vacateHearingRequest = vacateHearingRequest;
+        this.hearingRepository = hearingRepository;
+        this.hearingPartRepository = hearingPartRepository;
+        this.statusConfigService = statusConfigService;
+        this.statusServiceManager = statusServiceManager;
+        this.objectMapper = objectMapper;
+        this.entityManager = entityManager;
+    }
+
+    @Override
+    public void getAndValidateEntities() {
+        hearing = hearingRepository.findOne(vacateHearingRequest.getHearingId());
+        hearingParts = hearing.getHearingParts().stream()
+            .filter(hp -> statusServiceManager.canBeVacated(hp))
+            .collect(Collectors.toList());
+        sessions = hearingParts.stream()
+            .map(HearingPart::getSession)
+            .filter(Objects::nonNull)
+            .collect(Collectors.toList());
+        // Validation moved to act() due to conflict with optimistic lock
+
+        if (!statusServiceManager.canBeVacated(hearing)) {
+            throw new SnlEventsException("Hearing can not be vacated");
+        }
+    }
+
+    @Override
+    public UUID[] getAssociatedEntitiesIds() {
+        val ids = hearingParts.stream().map(HearingPart::getId).collect(Collectors.toList());
+        ids.addAll(hearingParts.stream()
+            .map(HearingPart::getSessionId)
+            .filter(Objects::nonNull)
+            .collect(Collectors.toList()));
+        ids.add(vacateHearingRequest.getHearingId());
+
+        return ids.stream().toArray(UUID[]::new);
+    }
+
+    @Override
+    public void act() {
+        try {
+            previousHearing = objectMapper.writeValueAsString(hearing);
+        } catch (JsonProcessingException e) {
+            throw new SnlRuntimeException(e);
+        }
+
+        entityManager.detach(hearing);
+        hearing.setVersion(vacateHearingRequest.getHearingVersion());
+        hearingRepository.save(hearing);
+
+        originalHearingParts = utdps.mapHearingPartsToStrings(objectMapper, hearingParts);
+
+        hearingParts.stream().forEach(hp -> {
+            hp.setStatus(statusConfigService.getStatusConfig(Status.Vacated));
```

Comment:
should hp.sessionId and hp.start not be set to null here to?

Reply:
now *setSession()* setter also sets *sessionId*. 
Well currently we do nothing with vacated hps so it wont bring any value, but for orderliness of data I will add it :)